### ISSUE TITLE
N3980: a more flexible way of hashing

### ISSUE THREAD

**Issue by amallia** [mavam/libbf#27]
The next version should have a more flexible way of hashing according to [N3980](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3980.html), which @mavam started in https://github.com/mavam/libbf/tree/topic/n3980. 

This issue will help us discussing about it.

**Comment by mavam** [comment]
Thanks for opening this one up. I've actually completed the heavy lifting in a separate project:

    https://github.com/vast-io/vast/tree/master/libvast/vast/concept/hashable

What remains is adjusting the *hasher* API.

**Comment by amallia** [comment]
Oh, this is cool!
This will allow to have any type hashable and so able to add to a BF. 
From my understand currently only arithmetic types (and arrays or vector of them) and string are supported.

**Comment by mavam** [comment]
> This will allow to have any type hashable and so able to add to a BF.

Exactly! It also decouples the hash algorithm implementation, so that we can use a variety of different algorithms in the future.
